### ISSUE TITLE
Improve accessibility with ARIA

### ISSUE THREAD

**Issue by desandro** [metafizzy/flickity#66]
Flickity could improve its accessibility by adding ARIA standards. We could use help here. Happy to consider your suggestions for adding ARIA attributes and roles.

http://www.w3.org/TR/wai-aria/

http://www.w3.org/WAI/tutorials/carousels/

http://accessibility.athena-ict.com/aria/examples/carousel.shtml

---

From what I gather, best route would be to recommend adding ARIA attributes into user's markup, rather than relying on Flickity to do it.

`aria-hidden` might be useful to highlight which cells are outside the carousel, but this would require adding logic to detect when cells are and are not visible.

**Comment by esseb** [comment]
> From what I gather, best route would be to recommend adding ARIA attributes into user's markup, rather than relying on Flickity to 

This really should be something the library does automatically. If the library supports this automatically you get free accessible carousels on any site that uses it. If you don't, most sites won't bother to do any work at all.

Plus, for `<li class="dot is-selected"></li>` I would want the dot to be an actually `<button>` so that it can be focused and activated using both enter and space. `focus()` should be called when clicking the dots (but not the previous and next buttons), and ideally a live region should announce which slide is currently visible (even if it's just "slide 1 of 3".) This requires deeper changes than I can do from outside the library.

**Comment by esseb** [comment]
Another accessible carousel example from http://examples.simplyaccessible.com/: http://examples.simplyaccessible.com/gazer/carousel.html

There seems little consensus on how to properly make a carousel accessible... I'll send some pull requests if I can figure anything out here.

**Comment by desandro** [comment]
[v1.2.0](https://github.com/metafizzy/flickity/releases/tag/v1.2.0) adds `aria-label` to previous and next buttons

**Comment by 2474** [comment]
> aria-hidden might be useful to highlight which cells are outside the carousel, but this would require adding logic to detect when cells are and are not visible.

Are you doing that logic already when you are determining what cell gets `.is-selected`? I'd think at that point you could set `aria-hidden: false` on that particular cell while setting `aria-hidden: true` to the rest.

Like you are possibly suggesting, if you leave ARIA attributes and roles to the user, do you think changing the `accessibility` to something different would be a good idea? When I think accessibility one of the first things that comes to mind is ARIA and potentially a user could misunderstand that option thinking that is what they'd be getting (in addition of keyboard nav).

Thanks for reading!

**Comment by 2474** [comment]
Hi @desandro, I took a stab at integrating, [	1ddf210](https://github.com/metafizzy/flickity/pull/350/commits/1ddf21025fbbece153114e6d5d458805812333a5) the aria hidden into Flickity by default, even if accessibility is toggled false. I still think there is value in having aria-hidden regardless as it really helps screen reader users and I don't think it adds much if any overhead. Let me know if I can improve anyway. Thanks!

**Comment by desandro** [comment]
> Are you doing that logic already when you are determining what cell gets .is-selected?

Only for the selected cell. But other cells may be visible if multiple cells fit in the carousel viewport, like with [various-width cells demo](http://codepen.io/desandro/pen/raJNdm), or [cellAlign left/right](http://flickity.metafizzy.co/options.html#cellalign). It's going to require more code to figure out what is and is not visible in the viewport.

Perhaps a better compromise is to set `aria-hidden` to `true` only for the selected cell, and not worry about other cells, even if they are visible. What do you think?

**Comment by 2474** [comment]
That is true, there would be cases where multiple cells might be visible but I would think a screen reader user wouldn't necessarily know that (unless they were only partially visually impaired, then they could see that). So I thought perhaps you could only have `aria-hidden="false"` on the selected and they would just work under the assumption it was like a normal slider (if that makes any sense).

Should I continue to just plug away with what I'm working on and perhaps when you are ready to implement this you could take a look?

Thanks for the quick response and insight.

**Comment by desandro** [comment]
As for Flickity v2.1.0, Flickity has ARIA attributes working:

+ Cell elements get `aria-selected` `true` or `false`
+ Page dots have `aria-label="Page dot #"`. Selected page dot gets `aria-current`
+ Previous & next buttons have `aria-label="Previous/Next"`

There's likely more improvement to be made. But I'm happy with the progress. I'm now closing this issue. Please open a new one to discuss other ARIA improvements.

**Comment by danijelGombac** [comment]
Are you considering to add the options for change aria-label values?

**Comment by desandro** [comment]
@danijelGombac Nope. The default labels are in english. If you do need to change them, you'll have to overwrite them with your own JavaScript.

**Comment by danijelGombac** [comment]
@desandro the problem is that you need to have different Javascript for each language, instead of just change options value.
